### ISSUE TITLE
write EPIPE error on windows when using --parallel option

### ISSUE THREAD

**Issue by vidz2** [cucumber/cucumber-js#1087]
I have been trying to get the experimental parallel option to work but I keep getting the below error on windows:
OS: windows 10, node: v8.9.4, cucumber: 4.2.1
```
Error: write EPIPE
    at _errnoException (util.js:1022:11)
    at Socket._writeGeneric (net.js:767:25)
    at Socket._write (net.js:786:8)
    at doWrite (_stream_writable.js:387:12)
    at writeOrBuffer (_stream_writable.js:373:5)
    at Socket.Writable.write (_stream_writable.js:290:11)
    at Socket.write (net.js:704:40)
    at Master.startSlave (C:\Users\bhargv\Documents\cucumber-demo\node_modules\cucumber\lib\runtime\parallel\master.js:108:27)
    at C:\Users\bhargv\Documents\cucumber-demo\node_modules\cucumber\lib\runtime\parallel\master.js:142:23
    at baseTimes (C:\Users\bhargv\Documents\cucumber-demo\node_modules\lodash\lodash.js:976:23)
    at Function.times (C:\Users\bhargv\Documents\cucumber-demo\node_modules\lodash\lodash.js:16116:20)
    at Master.run (C:\Users\bhargv\Documents\cucumber-demo\node_modules\cucumber\lib\runtime\parallel\master.js:141:24)
    at C:\Users\bhargv\Documents\cucumber-demo\node_modules\cucumber\lib\cli\index.js:207:35
    at Promise._execute (C:\Users\bhargv\Documents\cucumber-demo\node_modules\bluebird\js\release\debuggability.js:303:9)
    at Promise._resolveFromExecutor (C:\Users\bhargv\Documents\cucumber-demo\node_modules\bluebird\js\release\promise.js:483:18)
    at new Promise (C:\Users\bhargv\Documents\cucumber-demo\node_modules\bluebird\js\release\promise.js:79:10)
```

I get the above error on the nodejs example as well so its easy to reproduce. https://github.com/cucumber/cucumber-js/blob/master/docs/nodejs_example.md

( run command from cmd : node node_modules\cucumber\bin\cucumber-js --parallel 2 )

[cucumber-demo.zip](https://github.com/cucumber/cucumber-js/files/1972755/cucumber-demo.zip)

**Comment by lenntt** [comment]
Get a EPIPE error too when running on a mac if the tests log via `console`, likely related, although I don't see console statements in the attached zip.

**Comment by kyledmoore** [comment]
Ran into this myself and made the following project to illustrate, but didn't get around to logging it:
https://github.com/kyledmoore/cucumber-js-parallel-bug

I also noticed on my Mac that there was no improvement in run times when I added processes over running sequentially, suggesting that something might be arry w/ the asynchronous logic of the runner.    Not completely certain that my means for testing this was correct (I just passed the callback for a step definition to a setTimeout function):

`Given('something', function(cb){
    setTimeout(cb, 1000);
});`

https://github.com/kyledmoore/cucumber-js-parallel-bug/blob/master/features/step-definitions/steps.js

And had four features with scenarios containing that step.  I expected ~4second run with a single process, 1 second with 4.  Just occurred to me that I haven't tested multiple scenarios in a single feature.

**Comment by znevrly** [comment]
I have exactly the same error. Any update, please?

**Comment by hurrikam** [comment]
@charlierudolph could you please help with this? I am seeing the same issue as the guys above on Windows 10.

**Comment by hurrikam** [comment]
After a bit of debugging, I found what the issue is (I assume it's just on Windows).
In `src/runtime/parallel/master.js` at line 58 the slave process is wrongly initialized:

`const slaveProcess = childProcess.spawn(slaveCommand, [], {`

while it should be

`const slaveProcess = childProcess.spawn('node', [slaveCommand], {`

the `slaveCommand` variable initialized at the beginning of the file is just the absolute path to the `bin/run_slave` file, which is not an executable but rather just the argument to pass to the child node instance being created.

I applied and rebuilt the change on my machine and tested with two scenarios in two different features which just wait for a certain amount of time.
When running with `--parallel 2` the amount of time seemed to be about the longest of the two scenarios. Without parallel flag about the sum of them.

Note that the total time reported at the end of the suite run is still calculated as the sum of all the scenarios, even if in parallel mode it actually took much less.

@charlierudolph does the feature currently work on UNIX systems?

**Comment by pendenaor** [comment]
PR fix the issue, but any console.log raise a `SyntaxError` exception
```
Feature: Login failure

  @automated @sinoc @login
  Scenario: Try to login with invalid email address
    Given I open the `application page`

Feature: Login success

  @automated @sinoc @login
  Scenario: Try to login with valid credential
    Given I open the `application page`
    Then I see the `login form` `login input`
    Then I see the `login form` `login input`
    When I enter "filou" into `login form` `login input`
    When I enter "invalid-address" into `login form` `login input`
    And I enter "filoucfou" into `login form` `password input`
    And I enter "password" into `login form` `password input`
    And I click the `login form` `submit button`
    And I click the `login form` `submit button`
    Then I don't see a `login error` message
    Then I see a `login error` message
undefined:1
waitForElementMissingOrHidden result= { status: 0, value: true }
^

SyntaxError: Unexpected token w in JSON at position 0
    at JSON.parse (<anonymous>)
    at Master.parseSlaveLine (C:\Public\Repo\sinoc_ui-test\node_modules\cucumber\src\runtime\parallel\master.js:41:24)
    at Interface.<anonymous> (C:\Public\Repo\sinoc_ui-test\node_modules\cucumber\src\runtime\parallel\master.js:70:12)
    at emitOne (events.js:116:13)
    at Interface.emit (events.js:211:7)
    at Interface._onLine (readline.js:280:10)
    at Interface._normalWrite (readline.js:422:12)
    at Socket.ondata (readline.js:139:10)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:250:11)
    at Socket.Readable.push (_stream_readable.js:208:10)
    at Pipe.onread (net.js:597:20)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
```

**Comment by hurrikam** [comment]
@pendenaor that's the way parallelism works in Cucumber JS. The documentation warns indeed you can't use `console.log` as slaves communicates with the master process via stdout.
One solution would be to ignore output messages which are not in the format expected by the master process, but that's another discussion.

**Comment by pendenaor** [comment]
Ok, thanks for your explanation.

**Comment by charlierudolph** [comment]
Appears this is a bit old. Communication between master and slave was moved away from stdout. Appears still had some issues with pipe. See #1164. Going to close this as a duplicate of that. Expecting fix to be released in v6.
